clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/spark/utils/RowToAvroConverter.java

```diff
@@ -0,0 +1,475 @@
+package com.linkedin.venice.hadoop.spark.utils;
+
+import com.linkedin.avroutil1.compatibility.AvroCompatibilityHelper;
+import com.linkedin.venice.utils.ByteUtils;
+import java.math.BigDecimal;
+import java.nio.ByteBuffer;
+import java.sql.Date;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.Period;
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import org.apache.avro.Conversions;
+import org.apache.avro.LogicalType;
+import org.apache.avro.LogicalTypes;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericData;
+import org.apache.avro.generic.GenericFixed;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.commons.lang3.Validate;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.types.ArrayType;
+import org.apache.spark.sql.types.BinaryType;
+import org.apache.spark.sql.types.BooleanType;
+import org.apache.spark.sql.types.ByteType;
+import org.apache.spark.sql.types.DataType;
+import org.apache.spark.sql.types.DateType;
+import org.apache.spark.sql.types.DayTimeIntervalType;
+import org.apache.spark.sql.types.DecimalType;
+import org.apache.spark.sql.types.DoubleType;
+import org.apache.spark.sql.types.FloatType;
+import org.apache.spark.sql.types.IntegerType;
+import org.apache.spark.sql.types.LongType;
+import org.apache.spark.sql.types.MapType;
+import org.apache.spark.sql.types.ShortType;
+import org.apache.spark.sql.types.StringType;
+import org.apache.spark.sql.types.StructField;
+import org.apache.spark.sql.types.StructType;
+import org.apache.spark.sql.types.TimestampNTZType;
+import org.apache.spark.sql.types.TimestampType;
+import org.apache.spark.sql.types.YearMonthIntervalType;
+
+
+/**
+ * A utility class to convert Spark SQL Row to an Avro GenericRecord with the specified schema. This has been written in
+ * accordance with the following resources:
+ * <ul>
+ *   <li><a href="https://spark.apache.org/docs/latest/sql-data-sources-avro.html">Spark Avro data source documentation</a></li>
+ *   <li><a href="https://avro.apache.org/docs/1.11.1/spec.html">Avro specification</a></li>
+ *   <li><a href="https://github.com/apache/spark/blob/master/connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala">Spark's internal Catalyst type to Avro bytes implementation</a></li>
+ * </ul>
+ *
+ * Spark's implementation is not ideal to be used directly for two reasons:
+ * <ul>
+ *   <li>It cannot handle complex unions in the version of Spark that we use (3.1.1). The support was added in 3.4.0.</li>
```

Comment:
And I suppose there are reasons why we cannot upgrade to 3.4 yet?

Reply:
We use 3.1.1 because that's what we base the internal version at LI off of. When adding Spark support, I wanted to minimize the variables. I can try to bump to newer versions, but the transitive dependencies and stuff might not be happy. However, it is a bullet that we might need to bite

Reply:
Bah, it's fine to keep it at 3.1.1 for now I guess.